docs/writing-rules/data-flow/taint-mode.md

```diff
@@ -126,18 +152,300 @@ if cond:
+  at-exit: true
+```
+
+The above sink pattern matches either `return` statements (which are always "exit" statements), or function calls occuring as "exit" statements.
+
+Unlike regular sinks, at-exit sinks trigger a finding if any tainted l-value reaches the location of the sink. For example, the at-exit sink specification above will trigger a finding at a `return 0` statement if some tainted l-value reaches the `return`, even if `return 0` itself is not tainted. The location itself is the sink rather than the code that is at that location.
+
+We can use this, for example, to check that file descriptors are being closed within the same function where they were opened.
+
+<iframe src="https://semgrep.dev/embed/editor?snippet=OrAzB" border="0" frameBorder="0" width="100%" height="432"></iframe>
+
+The `print(content)` statement is reported because the control-flow exits the function at that point, and the file has not been closed.
+
+## Propagators (Pro)
+
+**Custom taint propagators is a Semgrep Pro feature.**
 
 By default, tainted data automatically propagates through assignments, operators, and function calls (from inputs to output). However, there are other ways in which taint can propagate, which can require language or library-specific knowledge that Semgrep does not have built-in.
 
+A taint propagator requires a pattern to be specified. Like in a search-mode rule, you can start this pattern with one of the following keys: `pattern`, `patterns`, `pattern-either`, `pattern-regex`.
+
+A propagator also needs to specify the origin (`from`) and the destination (`to`) of the taint to be propagated.
+
+| Field      | Type                       | Description                                                            |
+| :----------|:------------------------- | :--------------------------------------------------------------------- |
+| `from`     | metavariable              | Source of propagation. |
+| `to`       | metavariable              | Destination of propagation. |
+
+In addition, taint propagators accept the following options:
+
+| Option            | Type                      | Default | Description                                                            |
+| :-----------------|:------------------------- | :------ | :--------------------------------------------------------------------- |
+| `by-side-effect`  | {`false`, `true`} | `true` | See [_Propagation without side-effect_](#propagation-without-side-effect).               |
+
+For example, given the following propagator, if taint goes into the second argument of `strcpy`, its first argument will get the same taint:
+
+```yaml
+pattern-propagators:
+- pattern: strcpy($DST, $SRC)
+  from: $SRC
+  to: $DST
+```
+
+:::info
+Taint propagators only work intra-procedurally, that is, within a function or method. You cannot use taint propagators to propagate taint across different functions/methods, that is the job of [inter-procedural analysis](#inter-procedural-analysis-pro).
```

Comment:
```suggestion
Taint propagators only work intra-procedurally, that is, within a function or method. You cannot use taint propagators to propagate taint across different functions/methods. Use [inter-procedural analysis](#inter-procedural-analysis-pro).
```
Just confirming: does this mean to say that if a user flips on Semgrep Pro Engine, their taint rules that use propagators will all be able to detect interfile propagations? Can we say something like: turn on Semgrep Pro Engine so that your rules with propagators can perform interfile analysis.